### PR TITLE
napi: Added global support to ComponentInstance

### PR DESCRIPTION
**Pull request by FloVanGH**

### CODE REVIEW

File: api/napi/__test__/js_value_conversion_global.spec.ts [hunk 1]
```diff
@@ -0,0 +1,409 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-1.1 OR LicenseRef-Slint-commercial
+
+import test from 'ava';
+const path = require('node:path');
+var Jimp = require("jimp");
+
+import { ComponentCompiler, Brush, Model, Color, ImageData } from '../index'
+
+test('get/set string properties', (t) => {
```
Comment (tronical) [inline api/napi/__test__/js_value_conversion_global.spec.ts:10]: I think duplicating _all_ tests from `js_value_conversion` is overkill. That'll be a lot of extra work to maintain for no additional test coverage benefits. `js_value_conversion.spec.ts` is - as the name says - about testing conversion between JavaScript and Slint properties.

The tests here should not be about value conversion (that's already covered), it should be about property and callback access to globals. So I suggest to create one test that accesses say a string or number property of a global, tests that exceptions are thrown when getting the name of the global wrong, checks the error message when setting a non-existent property on an existing global, etc.. And then do the same with a second test about callbacks.

**Review by tronical — CHANGES_REQUESTED** [review]
Implementation looks good, but please simplify the tests to focus on the global property and callback access (and the error messages produced), instead of duplicating the image loading and value conversion tests (that might go out of sync the moment they are extended, for example with more model tests).

File: api/napi/__test__/js_value_conversion_global.spec.ts (hunk 1 above)
Comment (FloVanGH) [reply api/napi/__test__/js_value_conversion_global.spec.ts]: done.

File: api/napi/__test__/globals.spec.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-1.1 OR LicenseRef-Slint-commercial
+
+import test from 'ava';
+const path = require('node:path');
+var Jimp = require("jimp");
```
Comment (tronical) [inline api/napi/__test__/globals.spec.ts:6]:
```suggestion
```

File: api/napi/__test__/globals.spec.ts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+import { ComponentCompiler, Brush, Model, Color, ImageData } from '../index'
```
Comment (tronical) [inline api/napi/__test__/globals.spec.ts:8]:
```suggestion
import { ComponentCompiler } from '../index'
```

File: api/napi/__test__/globals.spec.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  t.not(definition, null);
+
+  let instance = definition!.create();
+  t.not(instance, null);
+  let speakTest;
+
+  t.throws(() => {
+    instance!.setGlobalCallback("MyGlobal", "great", (a: string, b: string, c: string, d: string, e: string) => {
+      speakTest = "hello " + a + ", " + b + ", " + c + ", " + d + " and " + e;
+    })
```
Comment (tronical) [inline api/napi/__test__/globals.spec.ts:98]:
```suggestion
    instance!.setGlobalCallback("MyGlobal", "great", () => {})
```

File: api/napi/__test__/globals.spec.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  },
+    {
+      code: "GenericFailure",
+      message: "Global MyGlobal not found"
+    }
+  );
+
+  t.throws(() => {
+    instance!.invokeGlobal("MyGlobal", "great", ["simon", "olivier", "auri", "tobias", "florian"])
```
Comment (tronical) [inline api/napi/__test__/globals.spec.ts:107]:
```suggestion
    instance!.invokeGlobal("MyGlobal", "great", [])
```

File: api/napi/__test__/globals.spec.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      "string"
+    }
+  }
+  export component App {}
+  `, "");
+  t.not(definition, null);
+
+  let instance = definition!.create();
+  t.not(instance, null);
+  let speakTest;
```
Comment (tronical) [inline api/napi/__test__/globals.spec.ts:93]:
```suggestion
```

File: api/napi/__test__/globals.spec.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+  t.throws(() => {
+    instance!.invokeGlobal("MyGlobal", "great", ["simon", "olivier", "auri", "tobias", "florian"])
+  },
+    {
+      code: "GenericFailure",
+      message: "Global MyGlobal not found"
+    }
+  );
+
```
Comment (tronical) [inline api/napi/__test__/globals.spec.ts:114]:
```suggestion

  let speakTest;
```

File: api/napi/src/interpreter/component_instance.rs [hunk 8]
```diff
@@ -65,6 +65,55 @@ impl JsComponentInstance {
...
+        let ty = self
+            .inner
+            .definition()
+            .global_properties_and_callbacks(global_name.as_str())
+            .ok_or(napi::Error::from_reason(format!("Global {global_name} not found")))?
+            .find_map(|(name, proptype)| if name == prop_name { Some(proptype) } else { None })
+            .ok_or(())
+            .map_err(|_| {
+                napi::Error::from_reason(format!(
+                    "Property {prop_name} of global {global_name} not found in the component"
```
Comment (tronical) [inline api/napi/src/interpreter/component_instance.rs:102]: Please add a test that covers this code path (you can fold that into the existing test).

File: api/napi/src/interpreter/component_instance.rs [hunk 9]
```diff
@@ -118,6 +167,61 @@ impl JsComponentInstance {
...
+        let ty = self
+            .inner
+            .definition()
+            .global_properties_and_callbacks(global_name.as_str())
+            .ok_or(napi::Error::from_reason(format!("Global {global_name} not found")))?
+            .find_map(|(name, proptype)| if name == callback_name { Some(proptype) } else { None })
+            .ok_or(())
+            .map_err(|_| {
+                napi::Error::from_reason(format!(
+                    "Callback {callback_name} of global {global_name} not found in the component"
```
Comment (tronical) [inline api/napi/src/interpreter/component_instance.rs:189]: Please add a test that covers this code path (you can fold that into the existing test).

File: api/napi/src/interpreter/component_instance.rs [hunk 10]
```diff
@@ -168,6 +272,63 @@ impl JsComponentInstance {
...
+            .inner
+            .definition()
+            .global_properties_and_callbacks(global_name.as_str())
+            .ok_or(napi::Error::from_reason(format!("Global {global_name} not found")))?
+            .find_map(|(name, proptype)| if name == callback_name { Some(proptype) } else { None })
+            .ok_or(())
+            .map_err(|_| {
+                napi::Error::from_reason(
+                    format!(
+                        "Callback {} of global {global_name} not found in the component",
```
Comment (tronical) [inline api/napi/src/interpreter/component_instance.rs:294]: Please add a test that covers this code path (you can fold that into the existing test).

**Review by tronical — APPROVED** [review]

**Pull request merged**